src/transformers/models/qwen2_vl/modeling_qwen2_vl.py

```diff
@@ -0,0 +1,1695 @@
+        L, _ = x.shape
+        q, k, v = self.qkv(x).reshape(L, 3, self.num_heads, -1).permute(1, 0, 2, 3).unbind(0)
+        if rotary_pos_emb is not None:
+            q = apply_rotary_pos_emb_vision(q.unsqueeze(0), rotary_pos_emb).squeeze(0)
+            k = apply_rotary_pos_emb_vision(k.unsqueeze(0), rotary_pos_emb).squeeze(0)
+
+        max_seqlen = (cu_seqlens[1:] - cu_seqlens[:-1]).max().item()
+        x = flash_attn_varlen_func(q, k, v, cu_seqlens, cu_seqlens, max_seqlen, max_seqlen).reshape(L, -1)
+        x = self.proj(x)
+        return x
+
+
+class VisionSpdaAttention(nn.Module):
+    def __init__(self, dim: int, num_heads: int = 16) -> None:
+        super().__init__()
+        self.num_heads = num_heads
+        self.qkv = nn.Linear(dim, dim * 3, bias=True)
+        self.proj = nn.Linear(dim, dim)
+
+    def forward(self, x: torch.Tensor, cu_seqlens: torch.Tensor, rotary_pos_emb: torch.Tensor = None) -> torch.Tensor:
+        L, _ = x.shape
+        q, k, v = self.qkv(x).reshape(L, 3, self.num_heads, -1).permute(1, 0, 2, 3).unbind(0)
+        if rotary_pos_emb is not None:
+            q = apply_rotary_pos_emb_vision(q.unsqueeze(0), rotary_pos_emb).squeeze(0)
+            k = apply_rotary_pos_emb_vision(k.unsqueeze(0), rotary_pos_emb).squeeze(0)
+
+        q = q.transpose(0, 1)
+        k = k.transpose(0, 1)
+        v = v.transpose(0, 1)
+        attn_outputs = []
+        for i in range(1, len(cu_seqlens)):
+            attn_outputs.append(
+                F.scaled_dot_product_attention(
+                    q[:, cu_seqlens[i - 1] : cu_seqlens[i], :],
+                    k[:, cu_seqlens[i - 1] : cu_seqlens[i], :],
+                    v[:, cu_seqlens[i - 1] : cu_seqlens[i], :],
+                    dropout_p=0.0,
+                    is_causal=False,
+                )
+            )
+        x = torch.cat(attn_outputs, dim=1).transpose(0, 1).reshape(L, -1)
+        x = self.proj(x)
+        return x
+
+
+QWEN2_VL_VISION_ATTENTION_CLASSES = {
+    "eager": VisionSpdaAttention,
+    "flash_attention_2": VisionFlashAttention2,
+    "sdpa": VisionSpdaAttention,
+}
+
+
+class Qwen2VLVisionBlock(nn.Module):
+    def __init__(self, config, attn_implementation: str = "sdpa") -> None:
+        super().__init__()
+        self.norm1 = LayerNorm(config.embed_dim, eps=1e-6)
+        self.norm2 = LayerNorm(config.embed_dim, eps=1e-6)
+        mlp_hidden_dim = int(config.embed_dim * config.mlp_ratio)
+
+        self.attn = QWEN2_VL_VISION_ATTENTION_CLASSES[attn_implementation](
+            config.embed_dim, num_heads=config.num_heads
+        )
+        self.mlp = VisionMlp(dim=config.embed_dim, hidden_dim=mlp_hidden_dim, hidden_act=config.hidden_act)
+
+    def forward(self, x, cu_seqlens, rotary_pos_emb) -> torch.Tensor:
+        x = x + self.attn(self.norm1(x), cu_seqlens=cu_seqlens, rotary_pos_emb=rotary_pos_emb)
+        x = x + self.mlp(self.norm2(x))
+        return x
+
+
+class Qwen2VisionTransformer(PreTrainedModel):
```

Comment:
Cool, thanks! But usually models should have a `QwenVLPretrainedModel` first and inherit from it and we won't need class attributes as they are same for Vision and LM models. Overall looks good to me

Reply:
hi @zucchini-nlp , should we introduce a `Qwen2VisionTransformerPretrainedModel`, even though the `Qwen2VisionTransformerPretrainedModel` doesn't have a separate configuration class? Could we utilize attributes from an existing `config` as the configuration class, like: 
`class Qwen2VisionTransformerPretrainedModel(PreTrainedModel):
    config_class = Qwen2VLConfig.vision_config
    base_model_prefix = "model"
     supports_gradient_checkpointing = True
    _no_split_modules = ["Qwen2VLVisionBlock"]
    _supports_flash_attn_2 = True
    _supports_sdpa = True
`
or make the `Qwen2VisionTransformer` to inherit from `Qwen2VLPretrainedModel` like this: `Qwen2VisionTransformer(Qwen2VLPretrainedModel):`?

Reply:
Or the config_class is not necessarily required, as in:
```
class Qwen2VisionTransformerPretrainedModel(PreTrainedModel):
     supports_gradient_checkpointing = True
    _no_split_modules = ["Qwen2VLVisionBlock"]
    _supports_flash_attn_2 = True
    _supports_sdpa = True

```

Reply:
> or make the Qwen2VisionTransformer to inherit from Qwen2VLPretrainedModel like this: Qwen2VisionTransformer(Qwen2VLPretrainedModel):?


IMO this is okey, and we can overwrite the `config_class` and `_no_split_modules` in Qwen2VisionTransformer while leaving other attributes with default values